File: 0279-perfect-squares/0279-perfect-squares.cpp
class Solution {
public:
    int numSquares(int n) {
        int cnt=0;
        vector<int> dp(n+1, INT_MAX);
        dp[0]=0;
        for(int i=1; i<=sqrt(n)+1; i++){
            for(int j=i*i; j<=n; j++){
                dp[j]= min(dp[j-(i*i)]+1, dp[j]);
            }
        }
        return dp[n];
    }
};

File: 0279-perfect-squares/0279-perfect-squares.java
class Solution {
    public int helper(int idx, ArrayList<Integer> vals, int n, int[] dp){
        if(idx==vals.size()){
            if(n==0) return 0;
            else return 10001;
        }
        if(dp[n]!= 10001) return dp[n];
        int mini= 10001;
        if(n>= vals.get(idx)) mini = Math.min(mini, helper(idx, vals, n-vals.get(idx), dp))+1;
        mini = Math.min(mini, helper(idx+1, vals, n, dp));
        dp[n]= mini;
        return mini;
    }
    public int numSquares(int n) {
        ArrayList<Integer> vals = new ArrayList<>();
        for(int i=1; i<=Math.pow(n, 0.5); i++){
            vals.add(i*i);
        }
        int []dp = new int[n+1];
            for(int j=0; j<=n; j++) dp[j]=10001;
        return helper(0, vals, n, dp);
    }
}